File: backend/app/routers/assets.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.schemas.common import UnifiedResponse
from app.schemas.asset import (
    AssetFolderCreate,
    AssetFolderUpdate,
    AssetFolderOut,
    AssetCreate,
    AssetUpdate,
    AssetOut,
)
from app.deps import get_db, get_current_user
from app.crud import asset as crud

router = APIRouter(prefix="/api/assets", tags=["assets"])


# --- Folders ---

@router.get("/folders", response_model=UnifiedResponse[list[AssetFolderOut]])
async def list_folders(
    space_key: str = Query("personal"),
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    folders = await crud.get_folders(db, user.id, space_key)
    return UnifiedResponse(
        code=200,
        data=[AssetFolderOut.model_validate(f) for f in folders],
        msg="ok",
    )


@router.post("/folders", response_model=UnifiedResponse[AssetFolderOut])
async def create_folder(
    body: AssetFolderCreate,
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    folder = await crud.create_folder(db, user.id, body.name, body.space_key, body.parent_id)
    if not folder:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Folder with this name already exists at this level",
        )
    return UnifiedResponse(code=200, data=AssetFolderOut.model_validate(folder), msg="created")


@router.put("/folders/{folder_id}", response_model=UnifiedResponse[AssetFolderOut])
async def update_folder(
    folder_id: int,
    body: AssetFolderUpdate,
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    if not body.name:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Name is required")
    folder = await crud.update_folder(db, folder_id, body.name, user.id)
    if not folder:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Folder not found")
    return UnifiedResponse(code=200, data=AssetFolderOut.model_validate(folder), msg="updated")


@router.delete("/folders/{folder_id}", response_model=UnifiedResponse)
async def delete_folder(
    folder_id: int,
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    ok = await crud.delete_folder(db, folder_id, user.id)
    if not ok:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Folder not found")
    return UnifiedResponse(code=200, msg="deleted")


@router.put("/items/batch", response_model=UnifiedResponse)
async def update_assets_batch(
    body: dict,
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    """Batch update assets: { ids: [1,2,3], updates: { folder_id: 5, type: "character" } }"""
    ids = body.get("ids", [])
    updates = body.get("updates", {})
    if not ids or not updates:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="ids and updates required")
    count = await crud.update_assets_batch(db, ids, user.id, updates)
    return UnifiedResponse(code=200, data={"count": count}, msg="updated")


# --- Assets ---

@router.get("/items", response_model=UnifiedResponse[dict])
async def list_assets(
    folder_id: Optional[int] = Query(None),
    asset_type: Optional[str] = Query(None, alias="type"),
    search: Optional[str] = Query(None),
    space_key: Optional[str] = Query(None),
    skip: int = Query(0),
    limit: int = Query(200),
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    assets = await crud.get_assets(db, user.id, folder_id, asset_type, search, space_key=space_key, skip=skip, limit=limit)
    total = await crud.count_assets(db, user.id, folder_id, asset_type, search, space_key=space_key)
    return UnifiedResponse(
        code=200,
        data={
            "items": [AssetOut.model_validate(a).model_dump(mode="json") for a in assets],
            "total": total,
        },
        msg="ok",
    )


@router.get("/items/source-urls", response_model=UnifiedResponse[list[str]])
async def list_source_urls(
    space_key: str = Query("personal"),
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    urls = await crud.get_asset_source_urls(db, user.id, space_key)
    return UnifiedResponse(code=200, data=urls, msg="ok")


@router.get("/items/{asset_id}", response_model=UnifiedResponse[AssetOut])
async def get_asset(
    asset_id: int,
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    asset = await crud.get_asset(db, asset_id, user.id)
    if not asset:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Asset not found")
    return UnifiedResponse(code=200, data=AssetOut.model_validate(asset), msg="ok")


@router.post("/items", response_model=UnifiedResponse[AssetOut])
async def create_asset(
    body: AssetCreate,
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    asset = await crud.create_asset(
        db, user.id,
        name=body.name, type=body.type,
        width=body.width, height=body.height,
        description=body.description, tags=body.tags, extra_data=body.extra_data,
        folder_id=body.folder_id, space_key=body.space_key,
    )
    return UnifiedResponse(code=200, data=AssetOut.model_validate(asset), msg="created")


@router.post("/items/batch", response_model=UnifiedResponse[list[AssetOut]])
async def create_assets_batch(
    body: list[AssetCreate],
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    items = [
        dict(name=b.name, type=b.type, width=b.width, height=b.height,
             description=b.description, tags=b.tags, extra_data=b.extra_data, folder_id=b.folder_id)
        for b in body
    ]
    assets = await crud.create_assets_batch(db, user.id, items)
    return UnifiedResponse(
        code=200,
        data=[AssetOut.model_validate(a) for a in assets],
        msg="created",
    )


@router.put("/items/{asset_id}", response_model=UnifiedResponse[AssetOut])
async def update_asset(
    asset_id: int,
    body: AssetUpdate,
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    asset = await crud.update_asset(
        db, asset_id, user.id,
        name=body.name, type=body.type,
        width=body.width, height=body.height,
        description=body.description, tags=body.tags, folder_id=body.folder_id,
    )
    if not asset:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Asset not found")
    return UnifiedResponse(code=200, data=AssetOut.model_validate(asset), msg="updated")


@router.delete("/items/{asset_id}", response_model=UnifiedResponse)
async def delete_asset(
    asset_id: int,
    db: AsyncSession = Depends(get_db),
    user=Depends(get_current_user),
):
    ok = await crud.delete_asset(db, asset_id, user.id)
    if not ok:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Asset not found")
    return UnifiedResponse(code=200, msg="deleted")

File: server/http/routes/assets.ts
/**
 * 资产与文件夹路由。
 * 处理资产、文件夹的查询、创建、更新与批量操作，所有写操作返回最新计数快照。
 */
import { Hono } from "hono";
import { Prisma } from "@prisma/client";
import { authenticateRequest } from "@server/core/auth/middleware";
import {
  folderCreateSchema,
  folderUpdateSchema,
  assetCreateSchema,
  assetUpdateSchema,
  assetBatchCreateSchema,
  assetBatchUpdateSchema,
} from "@server/schemas/asset";
import {
  AssetOperationError,
  getFolders,
  createFolder,
  getFolder,
  updateFolder,
  deleteFolder,
  getAssets,
  getAsset,
  updateAsset,
  deleteAsset,
  createAssetsBatch,
  updateAssetsBatch,
  listSourceUrls,
  getAssetLibrarySummary,
} from "@server/crud/asset";
import { ok, failCode } from "@server/core/response";
import {
  addAssetRef,
  removeAssetRef,
} from "@server/services/storage/ref-manager";
import { extractHashFromUrl } from "@server/utils/extract-hashes";

const router = new Hono();

/** 将 CRUD 层业务错误统一转换为 HTTP 错误码，避免事务细节泄漏到路由层。 */
function handleAssetError(error: unknown): Response | null {
  if (isSourceUrlUniqueError(error)) {
    return failCode(409, "assets.duplicate_source_url");
  }
  if (!(error instanceof AssetOperationError)) return null;
  if (error.code === "duplicate_source_url") {
    return failCode(409, "assets.duplicate_source_url");
  }
  if (error.code === "uncategorized_folder_protected") {
    return failCode(403, "assets.uncategorized_folder_protected");
  }
  if (error.code === "folder_not_found") {
    return failCode(404, "assets.folder_not_found");
  }
  return failCode(404, "assets.asset_not_found");
}

/** 判断唯一约束冲突是否来自资产来源 URL；避免把目录唯一冲突误报为来源重复。 */
function isSourceUrlUniqueError(error: unknown): boolean {
  if (!(error instanceof Prisma.PrismaClientKnownRequestError) || error.code !== "P2002") {
    return false;
  }
  const target = error.meta?.target;
  const targetText = Array.isArray(target) ? target.join(":") : String(target ?? "");
  return targetText.includes("source_url");
}

// ════════ Folders ════════

router.get("/api/assets/folders", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  const scope = c.req.query("scope") ?? "personal";
  const folders = await getFolders(auth.user.id, scope);
  return c.json(ok(folders));
});

router.get("/api/assets/bootstrap", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  const scope = c.req.query("scope") ?? "personal";
  const summary = await getAssetLibrarySummary(auth.user.id, scope);
  return c.json(ok(summary));
});

router.post("/api/assets/folders", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  let body: unknown;
  try {
    body = await c.req.json();
  } catch {
    return failCode(400, "common.invalid_json");
  }

  const parsed = folderCreateSchema.safeParse(body);
  if (!parsed.success) return failCode(422, "common.invalid_request");

  try {
    const folder = await createFolder(auth.user.id, parsed.data);
    return c.json(ok(folder));
  } catch (error) {
    return handleAssetError(error) ?? failCode(500, "common.internal_error");
  }
});

router.get("/api/assets/folders/:id", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  const id = parseInt(c.req.param("id"), 10);
  if (isNaN(id)) return failCode(400, "assets.invalid_folder_id");

  const folder = await getFolder(auth.user.id, id);
  if (!folder) return failCode(404, "assets.folder_not_found");
  return c.json(ok(folder));
});

router.put("/api/assets/folders/:id", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  const id = parseInt(c.req.param("id"), 10);
  if (isNaN(id)) return failCode(400, "assets.invalid_folder_id");

  let body: unknown;
  try {
    body = await c.req.json();
  } catch {
    return failCode(400, "common.invalid_json");
  }

  const parsed = folderUpdateSchema.safeParse(body);
  if (!parsed.success) return failCode(422, "common.invalid_request");

  try {
    const folder = await updateFolder(auth.user.id, id, parsed.data.name);
    return c.json(ok(folder));
  } catch (error) {
    return handleAssetError(error) ?? failCode(500, "common.internal_error");
  }
});

router.delete("/api/assets/folders/:id", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  const id = parseInt(c.req.param("id"), 10);
  if (isNaN(id)) return failCode(400, "assets.invalid_folder_id");

  try {
    const result = await deleteFolder(auth.user.id, id);
    await Promise.all(
      result.sourceUrls.map((sourceUrl) => {
        const hash = extractHashFromUrl(sourceUrl);
        return hash ? removeAssetRef(hash, auth.user.id) : Promise.resolve();
      }),
    );
    return c.json(ok(result));
  } catch (error) {
    return handleAssetError(error) ?? failCode(500, "common.internal_error");
  }
});

// ════════ Items ════════

router.get("/api/assets/items", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  const folderIdRaw = c.req.query("folder_id");
  const skipRaw = c.req.query("skip");
  const limitRaw = c.req.query("limit");
  const folderId = folderIdRaw ? parseInt(folderIdRaw, 10) : undefined;
  if (folderId !== undefined && isNaN(folderId)) return failCode(400, "assets.invalid_folder_id");

  const result = await getAssets({
    userId: auth.user.id,
    folderId,
    type: c.req.query("type") ?? undefined,
    search: c.req.query("search") ?? undefined,
    scope: c.req.query("scope") ?? undefined,
    skip: skipRaw ? parseInt(skipRaw, 10) : undefined,
    limit: limitRaw ? parseInt(limitRaw, 10) : undefined,
  });

  return c.json(ok({ items: result.items, total: result.total }));
});

router.get("/api/assets/items/source-urls", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  const scope = c.req.query("scope") ?? "personal";
  const urls = await listSourceUrls(auth.user.id, scope);
  return c.json(ok(urls));
});

router.post("/api/assets/items", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  let body: unknown;
  try {
    body = await c.req.json();
  } catch {
    return failCode(400, "common.invalid_json");
  }

  const parsed = assetCreateSchema.safeParse(body);
  if (!parsed.success) return failCode(422, "common.invalid_request");

  try {
    const result = await createAssetsBatch([{ ...parsed.data, userId: auth.user.id }]);
    const hash = result.items[0].sourceUrl ? extractHashFromUrl(result.items[0].sourceUrl) : null;
    if (hash) await addAssetRef(hash, auth.user.id);
    return c.json(ok({ item: result.items[0], counters: result.counters }));
  } catch (error) {
    return handleAssetError(error) ?? failCode(500, "common.internal_error");
  }
});

router.get("/api/assets/items/:id", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  const id = parseInt(c.req.param("id"), 10);
  if (isNaN(id)) return failCode(400, "assets.invalid_asset_id");

  const item = await getAsset(auth.user.id, id);
  if (!item) return failCode(404, "assets.asset_not_found");
  return c.json(ok(item));
});

router.put("/api/assets/items/batch", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  let body: unknown;
  try {
    body = await c.req.json();
  } catch {
    return failCode(400, "common.invalid_json");
  }

  const parsed = assetBatchUpdateSchema.safeParse(body);
  if (!parsed.success) return failCode(422, "common.invalid_request");

  try {
    const result = await updateAssetsBatch(auth.user.id, parsed.data.ids, parsed.data.updates);
    return c.json(ok(result));
  } catch (error) {
    return handleAssetError(error) ?? failCode(500, "common.internal_error");
  }
});

router.put("/api/assets/items/:id", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  const id = parseInt(c.req.param("id"), 10);
  if (isNaN(id)) return failCode(400, "assets.invalid_asset_id");

  let body: unknown;
  try {
    body = await c.req.json();
  } catch {
    return failCode(400, "common.invalid_json");
  }

  const parsed = assetUpdateSchema.safeParse(body);
  if (!parsed.success) return failCode(422, "common.invalid_request");

  try {
    const result = await updateAsset(auth.user.id, id, parsed.data);
    return c.json(ok(result));
  } catch (error) {
    return handleAssetError(error) ?? failCode(500, "common.internal_error");
  }
});

router.delete("/api/assets/items/:id", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  const id = parseInt(c.req.param("id"), 10);
  if (isNaN(id)) return failCode(400, "assets.invalid_asset_id");

  try {
    const result = await deleteAsset(auth.user.id, id);
    const hash = result.item.sourceUrl ? extractHashFromUrl(result.item.sourceUrl) : null;
    if (hash) await removeAssetRef(hash, auth.user.id);
    return c.json(ok({ counters: result.counters }));
  } catch (error) {
    return handleAssetError(error) ?? failCode(500, "common.internal_error");
  }
});

router.post("/api/assets/items/batch", async (c) => {
  const request = c.req.raw;
  const auth = await authenticateRequest(request);
  if ("error" in auth) return auth.error;

  let body: unknown;
  try {
    body = await c.req.json();
  } catch {
    return failCode(400, "common.invalid_json");
  }

  const parsed = assetBatchCreateSchema.safeParse(body);
  if (!parsed.success) return failCode(422, "common.invalid_request");

  try {
    const result = await createAssetsBatch(
      parsed.data.map((item) => ({ ...item, userId: auth.user.id }))
    );

    await Promise.all(
      result.items.map((item) => {
        const hash = item.sourceUrl ? extractHashFromUrl(item.sourceUrl) : null;
        return hash ? addAssetRef(hash, auth.user.id) : Promise.resolve();
      })
    );
    return c.json(ok({ items: result.items, counters: result.counters }));
  } catch (error) {
    return handleAssetError(error) ?? failCode(500, "common.internal_error");
  }
});

export { router };
